modules/src/clients/ics07_tendermint/client_def.rs

```diff
@@ -193,116 +197,198 @@ impl ClientDef for TendermintClient {
-        _prefix: &CommitmentPrefix,
-        _proof: &CommitmentProofBytes,
-        _connection_id: Option<&ConnectionId>,
-        _expected_connection_end: &ConnectionEnd,
+        client_state: &Self::ClientState,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        connection_id: &ConnectionId,
+        expected_connection_end: &ConnectionEnd,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        let path = Path::Connections(connection_id.clone()).to_string();
+        let value = expected_connection_end.encode_vec().unwrap();
+        verify_membership(client_state, prefix, proof, root, path, value)
     }
 
     fn verify_channel_state(
         &self,
-        _client_state: &Self::ClientState,
-        _height: Height,
-        _prefix: &CommitmentPrefix,
-        _proof: &CommitmentProofBytes,
-        _port_id: &PortId,
-        _channel_id: &ChannelId,
-        _expected_channel_end: &ChannelEnd,
+        client_state: &Self::ClientState,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+        expected_channel_end: &ChannelEnd,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        let path = Path::ChannelEnds(port_id.clone(), channel_id.clone()).to_string();
+        let value = expected_channel_end.encode_vec().unwrap();
+        verify_membership(client_state, prefix, proof, root, path, value)
     }
 
     fn verify_client_full_state(
         &self,
-        _client_state: &Self::ClientState,
-        _height: Height,
-        _root: &CommitmentRoot,
-        _prefix: &CommitmentPrefix,
-        _client_id: &ClientId,
-        _proof: &CommitmentProofBytes,
-        _expected_client_state: &AnyClientState,
+        client_state: &Self::ClientState,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        client_id: &ClientId,
+        expected_client_state: &AnyClientState,
     ) -> Result<(), Ics02Error> {
-        unimplemented!()
+        let path = Path::ClientState(client_id.clone()).to_string();
+        let value = expected_client_state.encode_vec().unwrap();
+        verify_membership(client_state, prefix, proof, root, path, value)
     }
 
     fn verify_packet_data(
         &self,
-        _client_state: &Self::ClientState,
-        _height: Height,
-        _proof: &CommitmentProofBytes,
-        _port_id: &PortId,
-        _channel_id: &ChannelId,
-        _seq: &Sequence,
-        _data: String,
+        client_state: &Self::ClientState,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+        sequence: Sequence,
+        commitment: String,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        let path = Path::Commitments {
```

Comment:
All packet related verification functions are additionally missing the following checks ->
```
// assert that enough time has elapsed
assert(currentTimestamp() >= processedTime + delayPeriodTime)
// assert that enough blocks have elapsed
assert(currentHeight() >= processedHeight + delayPeriodBlocks)
```

Reply:
@hu55a1n1 Thank you for the catch-up!
I'm trying to add the checks. I'm wondering that too many arguments (for times and heights) are given to the verification functions.
I have an idea to add a function like `validate_time()` to `ics02_client::client_def::ClientDef`:
```rust
pub trait ClientDef: Clone {
...
    fn validate_time(
        &self,
        current_time: Timestamp,
        current_height: Height,
        processed_time: Timestamp,
        processed_height: Height,
        delay_period_time: Duration,
        delay_period_blocks: u64,
    ) -> Result<(), Error>;
}
```
But I have a concern. The check is required by Tendermint client. I'm not sure it should be added to the generalized trait.

Reply:
Good idea! 👍 But since this is Tendermint specific, I think it would be better to add this to [`impl ClientState`](https://github.com/informalsystems/ibc-rs/blob/6d54a5ac5597a3b08d2384c3849204c8fbf0de7f/modules/src/clients/ics07_tendermint/client_state.rs#L166). Also, `verify_delay_passed()` might be a more appropriate name because we're dealing with both time and height delays.

Reply:
Thanks! I added `verify_delay_passed()` to Tendermint ClientState.